src/runner_metrics.py

```diff
@@ -0,0 +1,198 @@
+#  Copyright 2023 Canonical Ltd.
+#  See LICENSE file for licensing details.
+
+"""Classes and function to extract the metrics from a shared filesystem."""
+
+import json
+import logging
+from json import JSONDecodeError
+from typing import Optional
+
+from pydantic import BaseModel, NonNegativeFloat, ValidationError
+
+import metrics
+import shared_fs
+from errors import CorruptMetricDataError, MetricFileSizeTooLargeError
+
+logger = logging.getLogger(__name__)
+
+FILE_SIZE_BYTES_LIMIT = 1024
+PRE_JOB_METRICS_FILE_NAME = "pre-job-metrics.json"
+POST_JOB_METRICS_FILE_NAME = "post-job-metrics.json"
+RUNNER_INSTALLED_TS_FILE_NAME = "runner-installed.timestamp"
+
+
+class PreJobMetrics(BaseModel):
+    """Metrics for the pre-job phase of a runner.
+
+    Args:
+        timestamp: The UNIX time stamp of the time at which the event was originally issued.
+        workflow: The workflow name.
+        workflow_run_id: The workflow run id.
+        repository: The repository name.
+        event: The github event.
+    """
+
+    timestamp: NonNegativeFloat
+    workflow: str
+    workflow_run_id: str
+    repository: str
+    event: str
+
+
+class PostJobMetrics(BaseModel):
+    """Metrics for the post-job phase of a runner.
+
+    Args:
+        timestamp: The UNIX time stamp of the time at which the event was originally issued.
+        status: The status of the job.
+    """
+
+    timestamp: NonNegativeFloat
+    status: str
+
+
+class RunnerMetrics(BaseModel):
+    """Metrics for a runner.
+
+    Args:
+        installed_timestamp: The UNIX time stamp of the time at which the runner was installed.
+        pre_job: The metrics for the pre-job phase.
+        post_job: The metrics for the post-job phase.
+    """
+
+    installed_timestamp: NonNegativeFloat
+    pre_job: PreJobMetrics
+    post_job: Optional[PostJobMetrics]
+
+
+def _inspect_file_sizes(fs: shared_fs.SharedFilesystem) -> None:
+    """Inspect the file sizes of the shared filesystem.
+
+    Args:
+        fs: The path to the shared filesystem for a specific runner.
+
+    Raises:
+        FileSizeTooLargeError: If one of the files is too large.
+    """
+    files = [
+        fs.path.joinpath(PRE_JOB_METRICS_FILE_NAME),
+        fs.path.joinpath(POST_JOB_METRICS_FILE_NAME),
+        fs.path.joinpath(RUNNER_INSTALLED_TS_FILE_NAME),
+    ]
+    for file in files:
+        if file.exists() and file.stat().st_size > FILE_SIZE_BYTES_LIMIT:
+            raise MetricFileSizeTooLargeError(
+                f"File {file} is too large. The limit is {FILE_SIZE_BYTES_LIMIT} bytes."
+            )
```

Comment:
These kind of functions are interesting on whether they should raise exceptions or return data for the caller to raise exceptions with. An exception is conceptually that something went wrong during the execution. For the `_inspect_file_sizes`, it isn't an error for the file size to be too large since it is just inspecting it, so it should just return perhaps a boolean whether the inspection passed or failed and the caller can decide what to do with the information

Reply:
Yes, the name of the method does not necessarily indicate that an error would be thrown. I have updated the code.